### PR TITLE
fix heap buffer overflow by limiting http chunk size

### PR DESCRIPTION
**Pull request by CheSema**
02403_big_http_chunk_size -- test which make clickhouse crash.
Client sends data with chunked mode. It is up to client what to write to the chunks headers. Client is able to write there 2^64 us chunks size. Clickhouse parse chunks size from stream safely.  Chunks size is passed to the BufferWithOwnMemory::Memory::resize method. Integer overflow happens when Memory calculates how many bytes need to allocate, when padding is added. Turns out that Memory decided to allocate a few bytes when size was huge. After that HTTPChunkedReadBuffer writes a data after allocated memory.

I suggest to bound maximum chunk size which clickhouse accept. So there is new option http_max_chunk_size in the Setting with default value ~~32KB~~ 100G. Whatever implementation of Memory is used, Clickhouse should check the user input and rejects suspicious requests.
Also the implementation of class Memory have to be correct when big numbers are passed. If it is impossible, to add padding, then it throws exception instead wrong result.

In case when client sends chunk size bigger that value server just closes the connection. Unfortunately a client sees some generic exception without proper error message and code in that and similar cases. But server writes log with that event correctly with error log level.

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
This bug was found and send through ClickHouse bug-bounty [program](https://github.com/ClickHouse/ClickHouse/issues/38986) by *kiojj*.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by nickitat** [comment]
could you pls provide more detailed description of how everything is happening in error case? we should use `size_t` for dealing with sizes and to overflow it you need a chunk of impractical size. so I just don't really get the problem's mechanics

File: src/Core/Settings.h [hunk 1]
```diff
@@ -280,6 +280,7 @@ static constexpr UInt64 operator""_GiB(unsigned long long value)
     M(UInt64, http_max_fields, 1000000, "Maximum number of fields in HTTP header", 0) \
     M(UInt64, http_max_field_name_size, 1048576, "Maximum length of field name in HTTP header", 0) \
     M(UInt64, http_max_field_value_size, 1048576, "Maximum length of field value in HTTP header", 0) \
+    M(UInt64, http_max_chunk_size, 32*1024*1024, "Maximum value of a chunk size in HTTP chunked transfer encoding", 0) \
```
Comment (nickitat) [inline src/Core/Settings.h:283]: imo this value is too restrictive. we want just to protect ourselves from overflow. so maybe pick something that "ought to be enough for everyone", like 100GB or SLT?

**Comment by alexey-milovidov** [comment]
And we should not crash even if the setting was set to an overly large value.
Let's add one additional low-level check.

File: src/IO/BufferWithOwnMemory.h [hunk 2]
```diff
@@ -108,24 +112,47 @@ struct Memory : boost::noncopyable, Allocator
         if (!(value % alignment))
             return value;
 
-        return (value + alignment - 1) / alignment * alignment;
+        // original expression is (value + alignment - 1) / alignment * alignment;
+
+        size_t res = 0;
+
+        if (common::addOverflow<size_t>(value, alignment - 1, res))
+            throw Exception("value is too big to apply alignment", ErrorCodes::ARGUMENT_OUT_OF_BOUND);
```
Comment (CheSema) [inline src/IO/BufferWithOwnMemory.h:120]: I've made this calculation save and covered it with tests.
But.
I think that code is redundant. We don't have to round size manually. All the work is done by Allocator. The allocator is responsible for ensuring that the result address is a multiple of the alignment. It is done with posix_memalign.
I want to try make implementation without function align at all. Left only padding.

**Comment by CheSema** [comment]
Stress test(undefined) -- https://github.com/ClickHouse/ClickHouse/issues/39857

File: src/IO/BufferWithOwnMemory.h (hunk 2 above)
Comment (nickitat) [reply src/IO/BufferWithOwnMemory.h]: I guess the point was that you may allocate 1 byte with alignment to 16 bytes and have capacity 16. if you later call resize for 2, 3, 4, ... , 16 bytes it won't lead to reallocation

**Comment by nickitat** [comment]
> Test status: OK

just in case let me ask, have you checked the queries for which perf tests reported slowdowns?

**Review by nickitat — APPROVED** [review]

File: src/IO/BufferWithOwnMemory.h (hunk 2 above)
Comment (CheSema) [reply src/IO/BufferWithOwnMemory.h]: I also think that the purpose was exactly that. It should be named round instead alignment.
I greped the code. Class Memory usually is using with default alignment=0. Except one (? may be several) case with alignment=pageSize. And this is special case when pages are pinned. Of course, I could be wrong, but at least I've really tried to check this and this is the result of it.

**Comment by CheSema** [comment]
> > Test status: OK
> 
> just in case let me ask, have you checked the queries for which perf tests reported slowdowns?

Thank for showing how to do it.
Neither + nor - aren't confirmed in local runs. I run it on my branch and on base revision of that branch.

**Comment by tavplubix** [comment]
@CheSema, new test is flaky: https://play.clickhouse.com/play?user=play#c2VsZWN0IAp0b1N0YXJ0T2ZIb3VyKGNoZWNrX3N0YXJ0X3RpbWUpIGFzIGQsCmNvdW50KCksICBncm91cFVuaXFBcnJheShwdWxsX3JlcXVlc3RfbnVtYmVyKSwgIGFueShyZXBvcnRfdXJsKQpmcm9tIGNoZWNrcyB3aGVyZSAnMjAyMi0wNi0wMScgPD0gY2hlY2tfc3RhcnRfdGltZSBhbmQgdGVzdF9uYW1lIGxpa2UgJyUwMjQwM19iaWdfaHR0cF9jaHVua19zaXplJScgYW5kIHRlc3Rfc3RhdHVzIGluICgnRkFJTCcsICdGTEFLWScpIGdyb3VwIGJ5IGQgb3JkZXIgYnkgZCBkZXNj

**Comment by vitlibar** [comment]
> @CheSema, new test is flaky: https://play.clickhouse.com/play?user=play#c2VsZWN0IAp0b1N0YXJ0T2ZIb3VyKGNoZWNrX3N0YXJ0X3RpbWUpIGFzIGQsCmNvdW50KCksICBncm91cFVuaXFBcnJheShwdWxsX3JlcXVlc3RfbnVtYmVyKSwgIGFueShyZXBvcnRfdXJsKQpmcm9tIGNoZWNrcyB3aGVyZSAnMjAyMi0wNi0wMScgPD0gY2hlY2tfc3RhcnRfdGltZSBhbmQgdGVzdF9uYW1lIGxpa2UgJyUwMjQwM19iaWdfaHR0cF9jaHVua19zaXplJScgYW5kIHRlc3Rfc3RhdHVzIGluICgnRkFJTCcsICdGTEFLWScpIGdyb3VwIGJ5IGQgb3JkZXIgYnkgZCBkZXNj

https://github.com/ClickHouse/ClickHouse/pull/41996

File: src/Core/Settings.h [hunk 3]
```diff
@@ -280,6 +280,7 @@ static constexpr UInt64 operator""_GiB(unsigned long long value)
     M(UInt64, http_max_fields, 1000000, "Maximum number of fields in HTTP header", 0) \
     M(UInt64, http_max_field_name_size, 1048576, "Maximum length of field name in HTTP header", 0) \
     M(UInt64, http_max_field_value_size, 1048576, "Maximum length of field value in HTTP header", 0) \
+    M(UInt64, http_max_chunk_size, 100_GiB, "Maximum value of a chunk size in HTTP chunked transfer encoding", 0) \
```
Comment (azat) [inline src/Core/Settings.h:283]: @CheSema What is the reason for such a huge default? It is chunks not the whole request after all.

Comment (CheSema) [reply src/Core/Settings.h:283]: I would not call it as default. It is limitation. When you exceeded it you get an error.

Comment (CheSema) [reply src/Core/Settings.h:283]: It easy could be set with less value if such big values might break some logic after.

Comment (azat) [reply src/Core/Settings.h:283]: >I would not call it as default. It is limitation. When you exceeded it you get an error.

My bad, I meant "default for max chunk size"

Comment (azat) [reply src/Core/Settings.h:283]: Not an issue, but just a bad "recommendation" for users who will see such a big value and may think that it is a good idea to use such huge chunks.

Actually I came to this patch while I was looking at https://github.com/ClickHouse/ClickHouse/issues/55148

>Such rough limitation just enough for protection from integer overflow cases when server tries to allocate size near max<int_64>.

This is not a limitation, user can override it anyway.

Comment (CheSema) [reply src/Core/Settings.h:283]: > This is not a limitation, user can override it anyway.

You are right. Maybe that setting should not exist.

Comment (alexey-milovidov) [reply src/Core/Settings.h:283]: This is ok; we have a lot of settings of this type - "factory limits"

Comment (azat) [reply src/Core/Settings.h:283]: OK, how about https://github.com/ClickHouse/ClickHouse/pull/60852

**Pull request merged**